glean-core/src/traits/event.rs

```diff
@@ -0,0 +1,47 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+
+use crate::event_database::RecordedEvent;
+
+/// A description for the `EventMetric` type.
+///
+/// When changing this trait, make sure all the operations are
+/// implemented in the related type in `../metrics/`.
+pub trait Event {
+    /// Records an event.
+    ///
+    /// # Arguments
+    ///
+    /// * `timestamp` - A monotonically increasing timestamp, in milliseconds.
+    ///   This must be provided since the actual recording of the event may
+    ///   happen some time later than the moment the event occurred.
+    /// * `extra` - A HashMap of (key, value) pairs. The key is an index into
+    ///   the metric's `allowed_extra_keys` vector where the key's string is
+    ///   looked up. If any key index is out of range, an error is reported and
+    ///   no event is recorded.
+    fn record<M: Into<Option<HashMap<i32, String>>>>(&self, timestamp: u64, extra: M);
```

Comment:
Heh, after a FOG review I'm actually tasked to file a bug to get this changed. Wouldn't it be great if we don't need to take i32 but can take the enum instead? :) 

I'm also not super happy that we need to expose the `timestamp` argument here, but we can approach that in a follow-up eventually

Reply:
Oh wait, that's a mistake, we don't need to expose `timestamp`.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1672179

Reply:
Unfortunately I think we have to expose timestamp since we need to providing timing from other processes somehow.

Reply:
re timestamp: we kinda do need that API.
For FOG's IPC we need to get the timestamp immediately, pass that through the IPC layer and then later record the full event.
What we could do though is having a special\* `record_with_timestamp` API on the RLB only, not on the trait

\* I already dislike the "special" API, but I don't see a way around it

Reply:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1672198